File: apps/desktop/src-tauri/src/commands/credentials.rs
//! The keychain's IPC surface — deliberately **write-only**.
//!
//! Store, delete, ask whether one exists. There is no `get_credential` command and there must never
//! be one: `services/credential_store.rs` explains why at length, but the short version is that the
//! frontend having no way to *read* a secret is the whole point of moving them here. Anything that
//! needs a token — the GitHub proxy, the AI provider — reads it in Rust and puts it on the wire
//! without it ever crossing the IPC boundary.
//!
//! The write commands exist because a secret has to get in somehow: a personal access token the user
//! pastes, or an AI API key typed into Settings, both arrive in the webview by definition. That
//! one-way trip is unavoidable; a way back out is not.

use crate::services::credential_store::{self, CredentialKind};

/// Stores (or replaces) a secret for one account. An empty `secret` clears it — see `set_secret`.
#[tauri::command]
pub fn store_credential(kind: String, id: String, secret: String) -> Result<(), String> {
    let kind = CredentialKind::parse(&kind)?;
    credential_store::set_secret(kind, &id, &secret)?;
    Ok(())
}

/// Removes an account's secret — what disconnecting an account does.
#[tauri::command]
pub fn delete_credential(kind: String, id: String) -> Result<(), String> {
    let kind = CredentialKind::parse(&kind)?;
    credential_store::delete_secret(kind, &id)?;
    Ok(())
}

/// Whether a secret is stored for this account.
///
/// Safe to expose where a read is not: the answer is a boolean, and the frontend genuinely needs it
/// — an account can be listed in `settings.json` with nothing behind it in the keychain (a keychain
/// item revoked by hand, a settings file copied to another machine), and the UI has to be able to
/// say "reconnect this account" instead of firing a request that will come back 401.
#[tauri::command]
pub fn has_credential(kind: String, id: String) -> Result<bool, String> {
    let kind = CredentialKind::parse(&kind)?;
    Ok(credential_store::has_secret(kind, &id)?)
}

File: apps/desktop/src/lib/tauri/credentials.ts
import { invoke } from './invoke'

/**
 * The keychain, as seen from the webview: **write-only**.
 *
 * There is no `getCredential` here because there is no `get_credential` command, and there must
 * never be one. Every secret the app holds — a GitHub/GitLab/Bitbucket token, the AI provider's API
 * key — lives in the OS keychain, and the code that uses one reads it in Rust and puts it straight
 * on the wire. The frontend names an account; it cannot obtain the credential behind it.
 *
 * Writing is still possible because a secret has to arrive somehow: a token the user pastes, an API
 * key they type. That one-way trip is unavoidable. A way back out is not.
 *
 * See `src-tauri/src/services/credential_store.rs` for the full reasoning, including why this is a
 * service of our own rather than the `keyring` Tauri plugin (the plugin would expose a read).
 */
export type CredentialKind = 'github' | 'gitlab' | 'bitbucket' | 'ai'

/** The single id every AI provider key is stored under — there is one configured provider. */
export const AI_CREDENTIAL_ID = 'provider'

/** Stores (or replaces) a secret. An empty `secret` clears it. */
export const storeCredential = (kind: CredentialKind, id: string, secret: string) =>
  invoke<void>('store_credential', { kind, id, secret })

export const deleteCredential = (kind: CredentialKind, id: string) =>
  invoke<void>('delete_credential', { kind, id })

/**
 * Whether a secret is stored for this account — the one question about a credential the frontend is
 * allowed to ask, because the answer is a boolean rather than the credential.
 *
 * Worth asking: an account can be listed in the settings with nothing behind it in the keychain (an
 * item revoked by hand, a settings file copied to another machine), and the UI should say
 * "reconnect this account" rather than fire a request that comes back 401.
 */
export const hasCredential = (kind: CredentialKind, id: string) =>
  invoke<boolean>('has_credential', { kind, id })
